### PR TITLE
Configuration Table using Spreadsheet

### PR DESCRIPTION
**Pull request by realthunder**
Forum thread https://forum.freecadweb.org/viewtopic.php?f=17&t=42183

### CODE REVIEW

**Comment by wwmayer** [comment]
Note to myself: skip 5aa0f28835 as this is already in master

**Comment by berndhahnebach** [comment]
Following a link to the branch on the CI-repository:

https://gitlab.com/berndhahnebach/FreeCAD/-/commits/PR_5060

The CI-status is available on the latest commit of the branch.
If there is no status available the PR should be rebased on latest master.
Check pipeline branches to see if your PR has been run by the CI.

https://gitlab.com/berndhahnebach/FreeCAD/-/pipelines?scope=branches

**Comment by chennes** [comment]
OK, time to make some progress on this one! I'm going to start by cherry-picking the stuff that's peripheral to the main thrust of this PR and that doesn't need much help merging (I'll try to leave some notes about other commits):

Obsolete: e3873ffbb7257a6dc1bac14edb1b2d6c489e9111
Cherry-picked badd5b40b29fb23399e68a8687ab44a6fb034378
Cherry-picked a91d422ef0debdb47b4e60f9c073c2920c70f285
Needs rework: 08ab1b7a2aecddc7ee9bff1dfa4c518e1993f984
Cherry-picked 21b40c07953501e71ccb2873eb77982399d80e97
Cherry-picked d72b678205246df84fe7511bd50e94ae49c01357
Cherry-picked f56877d7e6bcb090c0e1ea5adaa390e0dddeabaa

It's not clear if `AutoParseContent` is necessary anymore, now that we require formulae to begin with '='. If it is still needed, then commit 316508b59 needs to be reworked.

Cherry-picked fb391b70bdc3092e71c1e498431a78f0e2fdf123

I'm stopping here, I'll get these merged and @realthunder can rebase, then I'll resume where I left off.

**Comment by realthunder** [comment]
Rebase is done.

> It's not clear if `AutoParseContent` is necessary anymore, now that we require formulae to begin with '='. If it is still needed

Yeah, I've skipped that commit.

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_2862">pipeline status</a>for feature branch [PR_2862](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_2862). Pipeline [#399136853 ](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/399136853) was triggered at [d3f0145](https://github.com/FreeCAD/FreeCAD/pull/2862/commits/d3f0145d307f4888d83180eaf2d43e80f5b901e5). All CI branch [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by gbroques** [comment]
@chennes Thank you for helping review, and move along this PR.

@realthunder Of course many thanks for all your work on these changes!

 The configuration table feature is *very* exciting, and your efforts do not go unoticed or unappreciated! :)

**Comment by chennes** [comment]
Does anything else in this PR depend on the work in "Spreadsheet: support paste into cell range" (df8ccf7)? It would be best if that could be extracted to its own PR, there are some design discussions to be had there, and we have to work out how to reimplement @mwganson's fix from PR #5030, and possibly mine from PR #5049 (I couldn't verify because of the 5030 bug).

**Comment by realthunder** [comment]
Yes, please skip that commit. I'll make a new PR for that in the future.

**Comment by realthunder** [comment]
BTW, I saw that in the upstream you copied empty cells between the selected range. It would be more efficient to just add code in paste range to clear the cell, which is implemented in that commit.

**Comment by chennes** [comment]
I think the whole copy-and-paste needs a rethink, it does not behave the same as either Excel or Google Sheets, which are the two biggest reference points we have. I'm fine with some deviation, but I think we should discuss it on the forums. I made some patches to fix obvious bugs, but there is clearly more work required.

**Comment by chennes** [comment]
OK, I've now gotten the following commits cherry-picked and merged:
* App: fix crash on copy of Enumeration abf8c4c
* App/Gui: improve expression binding of PropertyEnumeration  … b05d665
* App/Gui: allow change dynamic property group and documentation  … 8c1b752
* App: add href() expression built-in function  … 2b7ed0a

Next up I'm hoping to do everything except "App/Gui: support Link refresh on change of originally linked configur…" (afacab2) all at once, leaving just that one big one for the end.

**Comment by realthunder** [comment]
I'll do a rebase this week. Got something at hand right now.

**Comment by realthunder** [comment]
@chennes Rebase done. Sorry for the delay.

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_2862">pipeline status</a> for feature branch [PR_2862](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_2862). Pipeline [417926679](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/417926679) was triggered at [f96c91e](https://github.com/FreeCAD/FreeCAD/pull/2862/commits/f96c91eba00c512f14b8a6392b91aa4eb1bf01fd). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

File: src/Mod/PartDesign/Gui/ViewProviderShapeBinder.cpp [hunk 1]
```diff
@@ -259,7 +289,8 @@ void ViewProviderSubShapeBinder::setupContextMenu(QMenu* menu, QObject* receiver
     act = menu->addAction(QObject::tr("Synchronize"), receiver, member);
     act->setData(QVariant((int)0));
     act = menu->addAction(QObject::tr("Select bound object"), receiver, member);
-    act->setData(QVariant((int)1));
+    act->setData(QVariant((int)0x81));
```
Comment (chennes) [inline src/Mod/PartDesign/Gui/ViewProviderShapeBinder.cpp:292]: Is this just a random number?

Comment (realthunder) [reply src/Mod/PartDesign/Gui/ViewProviderShapeBinder.cpp:292]: Yes. Just some big number to not conflict with those defined in `Gui::ViewProvider`.

Comment (chennes) [reply src/Mod/PartDesign/Gui/ViewProviderShapeBinder.cpp:292]: Thanks - I see now that it doesn't matter, all of the data elements have been replaced by enums in the master. Sorry for the noise...

**Comment by chennes** [comment]
OK, it took a while, but I did get all but "App/Gui: support Link refresh on change of originally linked configur..." (9d1066f) cherry-picked and tested. They are in PR #5267 -- I'll give folks a few days to test it then get it merged.

**Comment by donovaly** [comment]
I think a decision should be made soon if this goes in for 0.20 or not.

**Comment by chennes** [comment]
I believe it just needs a single additional rebase, which should leave just one (large) commit to review. That said, the commit in question includes some macro soup, if I recall correctly.

**Comment by realthunder** [comment]
@chennes I have created a new PR #6401 with the remaining commit. I have partially removed some of the new macros introduced in that commit. There are still a few, but those are expansion of existing macro based code generation. I'll remove the macros in Link entirely in a separate PR later.

**Comment by chennes** [comment]
Sounds great, I will try to get that one reviewed ASAP. I guess you can close this one, then?

**Comment by donovaly** [comment]
Closing this since there is now #6401

**Pull request closed without merging**